simapp/app_v2.go

```diff
@@ -281,8 +274,12 @@ func NewSimApp(
 
 	app.sm.RegisterStoreDecoders()
 
-	// initialize BaseApp
-	app.SetInitChainer(app.InitChainer)
+	// Set custom init chainer that registers the module versions in the upgrade module
+	// NOTE: this is not required for apps that do not need custom pre-init genesis logic
+	app.SetInitChainer(func(ctx sdk.Context, req abci.RequestInitChain) abci.ResponseInitChain {
+		app.UpgradeKeeper.SetModuleVersionMap(ctx, app.ModuleManager.GetVersionMap())
```

Comment:
What if `runtime` provides a type:
```go
type HasModuleVersions interface {
  GetVersionMap() VersionMap
}
```

Then the upgrade module could get this as an optional dependency and would be able to call this during its `InitGenesis` and we could avoid having to do this manually.

Reply:
Runtime could even expose the full `ModuleManager` or even the `App` (as long as we require `Build` to be called to setup `BaseApp`. But this would be the smallest dependency

Reply:
Does it make sense to add module-specific logic in runtime?